core/src/main/java/com/linecorp/armeria/client/ClientRequestContext.java

```diff
@@ -161,6 +165,83 @@ static ClientRequestContextBuilder builder(RpcRequest request, URI uri) {
         return new ClientRequestContextBuilder(request, uri);
     }
 
+    /**
+     * Returns the {@link ServiceRequestContext} whose {@link Service} invokes the {@link Client}
+     * {@link Request} which created this {@link ClientRequestContext}, or {@code null} if this client request
+     * was not made in the context of a server request.
+     */
+    @Nullable
+    ServiceRequestContext rootContext();
+
+    /**
+     * Returns the {@link Attribute} for the given {@link AttributeKey}. This method will never return
+     * {@code null}, but may return an {@link Attribute} which does not have a value set yet.
+     *
+     * <p>If the {@link Attribute} is not in this context, but {@link #rootContext()} has it,
+     * then this will return the {@link Attribute}.
+     * Please note that if you change the {@link Attribute#get() value} from {@link #rootContext()} using
+     * setter methods (e.g. {@link Attribute#set(Object)}), a new {@link Attribute} for the given
+     * {@link AttributeKey} is created in this {@link ClientRequestContext} and the value will be set to
+     * the {@link Attribute}.
```

Comment:
How about this:
```
Please note that any changes made to the {@link Attribute} returned by this method never
affects the {@link Attribute} owned by {@link #rootContext()}. For example:
<pre>{@code
ClientRequestContext ctx = ...;
assert ctx.rootContext().attr(KEY).get() == 42;
assert ctx.hasOwnAttr(KEY) == false;
assert ctx.attr(KEY).get() == 42;
// Overriding the root context attribute creates the client context's own attribute.
ctx.attr(KEY).set(4242);
assert ctx.hasOwnAttr(KEY) == true;
// root context attribute remains unaffected.
assert ctx.rootContext().attr(KEY).get() == 42;
}</pre>
```

Reply:
minor nit: How about changing the values, which are assigned to the attribute, to meaningful values to make sure that the values belongs to.
```
// I'm not sure this is good. 😅
42 -> "ROOT"
4242 -> "CLIENT"
```